crates/ruff_python_formatter/tests/fixtures.rs

```diff
@@ -239,6 +239,55 @@ Formatted twice:
     }
 }
 
+/// Ensure that formatting doesn't change the AST.
+///
+/// Like Black, there are a few exceptions to this "invariant" which are encoded in
+/// [`NormalizedMod`] and related structs. Namely, formatting can change indentation within strings,
+/// and can also flatten tuples within `del` statements.
+fn ensure_unchanged_ast(
+    unformatted_code: &str,
+    formatted_code: &str,
+    options: &PyFormatOptions,
+    input_path: &Path,
+) {
+    let source_type = options.source_type();
+
+    // Parse the unformatted code.
+    let unformatted_ast = parse(
+        unformatted_code,
+        source_type.as_mode(),
+        &input_path.to_string_lossy(),
+    )
+    .expect("Unformatted code to be valid syntax");
+    let unformatted_ast = NormalizedMod::from(&unformatted_ast);
+
+    // Parse the formatted code.
+    let formatted_ast = parse(
+        formatted_code,
+        source_type.as_mode(),
+        &input_path.to_string_lossy(),
+    )
+    .expect("Formatted code to be valid syntax");
+    let formatted_ast = NormalizedMod::from(&formatted_ast);
+
+    if formatted_ast != unformatted_ast {
+        let diff = TextDiff::from_lines(
+            &format!("{unformatted_ast:#?}"),
+            &format!("{formatted_ast:#?}"),
+        )
+        .unified_diff()
+        .header("Unformatted", "Formatted")
+        .to_string();
+        panic!(
+            r#"Reformatting the unformatted code of {} resulted in AST changes.
```

Comment:
```suggestion
            r#"Reformatting the unformatted code of {} resulted in AST changes. Please open an issue at https://github.com/astral-sh/ruff/issues/new with your configuration and the failing code snippet
```
or something along the lines

Reply:
I think it's fine to omit since this is in tests only.